### PR TITLE
Manage collaborators through ttnctl

### PR DESCRIPTION
**Pull request by gomezjdaniel**
Added commands under `applications`, `gateways`, `components`

- `ttnctl applications collaborators add [AppID] [Username] [Rights...]`
- `ttnctl applications collaborators delete [AppID] [Username]`
- `ttnctl gateways collaborators add [AppID] [Username] [Rights...]`
- `ttnctl gateways collaborators delete [AppID] [Username]`
- `ttnctl components collaborators add [Type] [ComponentID] [Username] [Rights...]`
- `ttnctl components collaborators delete [Type] [ComponentID] [Username]`

Also, `info` command prints the list of collaborators.

### CODE REVIEW

File: ttnctl/cmd/components_collaborators_delete.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	Example: `$ ttnctl components collaborators delete handler ttn-handler-dev gomezjdaniel
+  INFO Successfully user retracted as component collaborator ComponentID=dev-handler Type=handler Username=gomezjdaniel
+`,
+	Run: func(cmd *cobra.Command, args []string) {
+		assertArgsLength(cmd, args, 3, 3)
+
+		account := util.GetAccount(ctx)
+
+		if err := account.RetractComponentRights(args[0], args[1], args[2]); err != nil {
+			ctx.WithError(err).Fatal("Could not retract user as component collaborator")
```
Comment (johanstokking) [inline ttnctl/cmd/components_collaborators_delete.go]: Consistent naming; delete vs retract. Also see `Short` description. Use delete everywhere

File: ttnctl/cmd/components_collaborators_set.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	"github.com/TheThingsNetwork/ttn/core/types"
+	"github.com/TheThingsNetwork/ttn/ttnctl/util"
+	"github.com/spf13/cobra"
+)
+
+var componentsCollaboratorsSetCmd = &cobra.Command{
+	Use:   "set [Type] [ComponentID] [Username]",
+	Short: "Grant an user with rights over a network component.",
+	Long:  `components collaborators set can be used to add an username as collaborator or edit his rights.`,
+	Example: `$ ttnctl components collaborators set handler ttn-handler-dev gomezjdaniel --rights='component:settings,component:delete'
```
Comment (johanstokking) [inline ttnctl/cmd/components_collaborators_set.go]: Remove your own username

File: ttnctl/cmd/components_collaborators_delete.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	ttnlog "github.com/TheThingsNetwork/go-utils/log"
+	"github.com/TheThingsNetwork/ttn/ttnctl/util"
+	"github.com/spf13/cobra"
+)
+
+var componentsCollaboratorsDeleteCmd = &cobra.Command{
+	Use:   "delete [Type] [ComponentID] [Username]",
+	Short: "Retracts an user as component collaborator.",
+	Long:  `components collaborators delete can be used to retract users as collaborators.`,
+	Example: `$ ttnctl components collaborators delete handler ttn-handler-dev gomezjdaniel
```
Comment (johanstokking) [inline ttnctl/cmd/components_collaborators_delete.go]: Remove your own username

File: ttnctl/cmd/components_collaborators_set.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+		if err := account.GrantComponentRights(args[0], args[1], args[2], rights); err != nil {
+			ctx.WithError(err).Fatal("Could not grant user with collaborator rights")
+		}
+
+		ctx.WithFields(ttnlog.Fields{
+			"Type":        args[0],
+			"ComponentID": args[1],
+			"Username":    args[2],
+		}).Info("Successfully user granted with component rights")
```
Comment (johanstokking) [inline ttnctl/cmd/components_collaborators_set.go]: Inconsistent naming; set vs grant

File: ttnctl/cmd/components_collaborators_set.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	"github.com/TheThingsNetwork/ttn/ttnctl/util"
+	"github.com/spf13/cobra"
+)
+
+var componentsCollaboratorsSetCmd = &cobra.Command{
+	Use:   "set [Type] [ComponentID] [Username]",
+	Short: "Grant an user with rights over a network component.",
+	Long:  `components collaborators set can be used to add an username as collaborator or edit his rights.`,
+	Example: `$ ttnctl components collaborators set handler ttn-handler-dev gomezjdaniel --rights='component:settings,component:delete'
+  INFO Successfully user granted with collaborator rights ComponentID=ttn-dev-handler Type=handler Username=gomezjdaniel
```
Comment (johanstokking) [inline ttnctl/cmd/components_collaborators_set.go]: Remove your own username

File: ttnctl/cmd/components_collaborators_set.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+		assertArgsLength(cmd, args, 3, 3)
+
+		var rights []types.Right
+		rightsStrings, err := cmd.Flags().GetStringSlice("rights")
+		if err != nil {
+			ctx.WithError(err).Fatal("Could not read rights options")
+		}
+		if len(rightsStrings) == 0 {
+			ctx.Fatal("Rights has not been provided through --rights flag")
+		}
```
Comment (johanstokking) [inline ttnctl/cmd/components_collaborators_set.go]: Is the user supposed to know the rights? May be, but some help/doc/command description is required then

File: ttnctl/cmd/components_info.go [hunk 7]
```diff
@@ -38,6 +38,19 @@ Created:        2016-10-06 09:52:28.766 +0000 UTC
...
 		fmt.Println()
+		fmt.Println("Collaborators:")
+		for _, collaborator := range component.Collaborators {
+			fmt.Printf("       - Name: %s\n", collaborator.Username)
+			fmt.Print("         Rights: ")
+			for i, right := range collaborator.Rights {
+				if i != 0 {
+					fmt.Print(", ")
+				}
+				fmt.Print(right)
```
Comment (johanstokking) [inline ttnctl/cmd/components_info.go]: How about `strings.Join(collaborator.Rights, ",")`

Comment (htdvisser) [inline ttnctl/cmd/components_info.go]: Unfortunately `rights` is a `[]types.Right`, so that won't work

Comment (johanstokking) [inline ttnctl/cmd/components_info.go]: OK. Would have been nice to have had a `type Rights []Right` then. I can imagine many quick operations to check rights, add right, remove right, format rights, etc. Can add that to https://github.com/TheThingsNetwork/go-account-lib/issues/63. Anyway, leave it here for now

File: ttnctl/cmd/applications_collaborators_delete.go [hunk 8]
```diff
@@ -9,9 +9,10 @@ import (
 )
 
 var applicationsCollaboratorsDeleteCmd = &cobra.Command{
-	Use:   "delete [AppID] [Username]",
-	Short: "Delete collaborators from an application.",
-	Long:  `applications collaborators delete can be used to delete a collaborator from an application.`,
+	Use:     "delete [AppID] [Username]",
+	Aliases: []string{"remove"},
+	Short:   "Delete collaborators from an application.",
```
Comment (johanstokking) [inline ttnctl/cmd/applications_collaborators_delete.go]: While you're at it: `a collaborator` singular

**Comment by htdvisser** [comment]
Putting this on hold, as the Account Server is currently not handling the commands correctly. Waiting for response from @romeovs

File: ttnctl/cmd/components_info.go (hunk 7 above)
Comment (htdvisser) [inline ttnctl/cmd/components_info.go]: Agreed. Created `joinRights` for now

File: ttnctl/cmd/collaborators.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	"github.com/TheThingsNetwork/ttn/core/types"
+	"github.com/gosuri/uitable"
+)
+
+func printCollaborators(collaborators []account.Collaborator) {
+	table := uitable.New()
+	table.MaxColWidth = 70
+	table.AddRow("", "Username", "Rights")
+	for i, collaborator := range collaborators {
+		table.AddRow(i+1, collaborator.Username, joinRights(collaborator.Rights, ","))
```
Comment (romeovs) [inline ttnctl/cmd/collaborators.go]: Join with `", "`

File: ttnctl/cmd/components_info.go [hunk 10]
```diff
@@ -38,6 +38,11 @@ Created:        2016-10-06 09:52:28.766 +0000 UTC
 		fmt.Printf("Type:           %s\n", component.Type)
 		fmt.Printf("Created:        %s\n", component.Created)
 		fmt.Println()
```
Comment (romeovs) [inline ttnctl/cmd/components_info.go]: Remove this `Println` for consistent output.

File: ttnctl/cmd/gateways_info.go [hunk 11]
```diff
@@ -62,6 +62,11 @@ var gatewaysInfoCmd = &cobra.Command{
 		}
 
 		fmt.Println()
```
Comment (romeovs) [inline ttnctl/cmd/gateways_info.go]: Remove `Println`

**Comment by romeovs** [comment]
Account server is deployed with the required endpoints and I fixed an issue with incorrect rights.
Added some more change requests for @gomezjdaniel

**Pull request merged**